include/knowhere/version.h

```diff
@@ -0,0 +1,141 @@
+            patch_version = unexpected_version_num;
+        }
+    }
+
+    bool
+    Valid() {
+        return major_version != unexpected_version_num && minor_version != unexpected_version_num &&
+               patch_version != unexpected_version_num;
+    };
+
+    std::string
+    VersionCode() {
+        return version_code;
+    }
+
+    static bool
+    VersionCheck(std::string version) {
+        try {
+            return std::regex_match(version.c_str(), std::regex(version_regex));
+        } catch (std::regex_error& e) {
+            LOG_KNOWHERE_ERROR_ << "unexpected index version : " << version;
+        }
+        return false;
+    }
+
+    // used when version is not set
+    static inline Version
+    GetDefaultVersion() {
+        return Version(default_version);
+    }
+
+    // the current version (newest version support)
+    static inline Version
+    GetCurrentVersion() {
+        return Version(current_version);
+    }
+
+    // the minimal version (oldest version support)
+    static inline Version
+    GetMinimalSupport() {
+        return Version(minimal_vesion);
+    }
+
+    static inline bool
+    VersionSupport(Version version) {
+        return VersionCheck(version.version_code) && GetMinimalSupport() < version;
+    }
+
+    friend bool
+    operator<(const Version& lhs, const Version& rhs) {
+        if (lhs.major_version != rhs.major_version) {
+            return lhs.major_version < rhs.major_version;
+        } else if (lhs.minor_version != rhs.minor_version) {
+            return lhs.minor_version < rhs.minor_version;
+        } else if (lhs.patch_version != rhs.patch_version) {
+            return lhs.patch_version < rhs.patch_version;
+        }
+        return true;
+    }
+
+    friend bool
+    operator>(const Version& lhs, const Version& rhs) {
+        return operator<(rhs, lhs);
+    }
+
+ private:
+    static std::vector<int32_t>
+    version_split(std::string version_str) {
+        if (!VersionCheck(version_str)) {
+            return std::vector<int32_t>();
+        }
+        try {
+            auto prefix_pos = version_str.find(prefix_delimiter);
+            auto prefix = version_str.substr(0, prefix_pos);
+            auto suffix = version_str.substr(prefix_pos + 1);
+            auto num_str = suffix.substr(1);
+            auto right = num_str.find(code_delimiter);
+            std::vector<int32_t> version_nums;
+            while (right != std::string::npos) {
+                auto version_num = num_str.substr(0, right);
+                version_nums.emplace_back(atoi(version_num.c_str()));
+                num_str = num_str.substr(right + 1);
+                right = num_str.find(code_delimiter);
+            }
+            version_nums.emplace_back(atoi(num_str.c_str()));
+            return version_nums;
+        } catch (std::exception& e) {
+            LOG_KNOWHERE_ERROR_ << "unexpected version code : " << version_str;
+            return std::vector<int32_t>();
+        }
+    }
```

Comment:
this can be written using regex group matching which is much simpler. example:

```cpp
    std::regex pattern(R"(^(knowhere)-[vV](\d+)\.(\d+)\.(\d+)$)");
    std::smatch match;
    std::vector<int> parts;

    if (std::regex_match(version, match, pattern)) {
        parts.push_back(std::stoi(match[2].str()));  // major version
        parts.push_back(std::stoi(match[3].str()));  // minor version
        parts.push_back(std::stoi(match[4].str()));  // patch version
    }

    return parts;
```

Reply:
cool~